### DISCUSSION TITLE
record ToString Questions and suggestions

### DISCUSSION THREAD

**Discussion by bunnyi116** [dotnet/csharplang#7082] [General]
Is it a bug that I find that override ToString is invalid from the abstract record base type and can only be effective from Derived Class? This will override many Derived ToString methods if it cannot be overwritten from the abstract record, which is cumbersome for many Derived classes.

I have one more suggestion for record: Add an Attribute that ignores PrintMember, so instead of each override Derived Class override PrintMembers method, Simply add the specified Attribute on a particular attribute or field to ignore (similar to '[JsonIgnore]')

For the PrintMemberAttribute suggestion, the attributes can be: IgnorePrintMembers, IgnoreNull

If the compiler can do this automatically at compile time, you can subtract a lot of unnecessary work and time (which shouldn't be hard to do, because that's what the compiler or code generator does).

## Test the model

```C#
using System.Text;

var models = new List<BaseModel>() { new ModelA(), new ModelB(), };

foreach (var model in models)
{
    Console.WriteLine(model.ToString());
    Console.WriteLine(model.ToString2());
    Console.WriteLine();
}

enum ModelType { Unknown, ModelA, ModelB }
abstract record BaseModel
{
    public virtual string RawType { get; init; } = "Unknown";
    public ModelType Type => Enum.Parse<ModelType>(RawType);
    protected virtual bool PrintMembers(StringBuilder builder)
    {
        builder.Append($"RawType = {RawType}");
        return true;
    }
    public override string ToString() => "Override ToString！";
    public virtual string ToString2() => "ToString2！";
}

record ModelA : BaseModel
{
    public override string RawType { get; init; } = "ModelA";
    public string ExtraA { get; init; } = "ExtraA";
}

record ModelB : BaseModel
{
    public override string RawType { get; init; } = "ModelB";
    public string ExtraB { get; init; } = "ExtraB";
    public override string ToString() => "Override ModelB ToString！";
    public override string ToString2() => "Override ModelB ToString2！";
}

```

## Running results
```
ModelA { RawType = ModelA, ExtraA = ExtraA }
ToString2！

Override ModelB ToString！
Override ModelB ToString2！
```

**Comment by bunnyi116** [comment]
NET 7.0

**Comment by huoyaoyuan** [comment]
Every record overrides `ToString` based on its members, so the override from a base record will be overridden again.
You can use `sealed override` to prevent derived records to override it and force the behavior. However you will lose the right to override it again. In your sample, `ModelA` will get the `ToString` from `BaseModel`, and `ModelB` will get compilation error.

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-meta
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

Is it a bug that I find that override ToString is invalid from the abstract record base type and can only be effective from Derived Class? This will override many Derived ToString methods if it cannot be overwritten from the abstract record, which is cumbersome for many Derived classes.

I have one more suggestion for record: Add an Attribute that ignores PrintMember, so instead of each override Derived Class override PrintMembers method, Simply add the specified Attribute on a particular attribute or field to ignore (similar to '[JsonIgnore]')

For the PrintMemberAttribute suggestion, the attributes can be: IgnorePrintMembers, IgnoreNull

If the compiler can do this automatically at compile time, you can subtract a lot of unnecessary work and time (which shouldn't be hard to do, because that's what the compiler or code generator does).

## Test the model

```C#
using System.Text;

var models = new List<BaseModel>() { new ModelA(), new ModelB(), };

foreach (var model in models)
{
    Console.WriteLine(model.ToString());
    Console.WriteLine(model.ToString2());
    Console.WriteLine();
}

enum ModelType { Unknown, ModelA, ModelB }
abstract record BaseModel
{
    public virtual string RawType { get; init; } = "Unknown";
    public ModelType Type => Enum.Parse<ModelType>(RawType);
    protected virtual bool PrintMembers(StringBuilder builder)
    {
        builder.Append($"RawType = {RawType}");
        return true;
    }
    public override string ToString() => "Override ToString！";
    public virtual string ToString2() => "ToString2！";
}

record ModelA : BaseModel
{
    public override string RawType { get; init; } = "ModelA";
    public string ExtraA { get; init; } = "ExtraA";
}

record ModelB : BaseModel
{
    public override string RawType { get; init; } = "ModelB";
    public string ExtraB { get; init; } = "ExtraB";
    public override string ToString() => "Override ModelB ToString！";
    public override string ToString2() => "Override ModelB ToString2！";
}

```

## Running results
```
ModelA { RawType = ModelA, ExtraA = ExtraA }
ToString2！

Override ModelB ToString！
Override ModelB ToString2！
```

<table>
  <tr>
    <th align="left">Author:</th>
    <td>Bunnui</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`question`, `area-Meta`, `untriaged`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Comment by bunnyi116** [comment]
> Every record overrides `ToString` based on its members, so the override from a base record will be overridden again. You can use `sealed override` to prevent derived records to override it and force the behavior. However you will lose the right to override it again. In your sample, `ModelA` will get the `ToString` from `BaseModel`, and `ModelB` will get compilation error.

`ModelA` failed to get the `override string ToString()` from `BaseModel`. It used `record`s default `ToString()`, which means `BaseModel`, `override string ToString()` does not work.

I think we should check if `override string ToString()` exists when generating the source code, and if it does, instead of generating `ToString()`, we should use the user's `override string ToString()`, Just like `ToString2()` in `BaseModel`

**Comment by huoyaoyuan** [comment]
> `ModelA` failed to get the `override string ToString()` from `BaseModel`. It used `record`s default `ToString()`, which means `BaseModel`, `override string ToString()` does not work.

If you decompile the sample, you will see that `ModelA` overrides it again. There's not a "default" implementation for `record`. Every record just reimplement it in that way. See:
https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwCwG5lqZwB0AKgKYAeALvksgG4CGATugLYD2w5ANgM7oAvOgB25AO7oAMgEt+1ADwAhJv3IBZbnwB8ACgCU6AN5jJ6LT14BBQwBozUy32X30AXzrIAZpxbkmAGMAC3Q9ZjYuK3RZUQ5tAQNkY2R0NKIATj0ovjJOAGVqFliAc0MDOnTM7IS8wuLREphyyvSsLIrkd2RkclEAV3YLBNIATwAHchN0AFVRAGtRTglRB2cbNYTlD2QmACMFFiDqdH9Av2B0VXV15NS2gGZ0BlkWan6mXiIMACUmCTGk2mJXItBiolkYPcwnQACI5otlqJYa00lAnutAVMscIdOgAKIDdjEAAKrHUikxE3I+j+AOpnSQVXGLE41HIgXZlxebw+Xz2nE4XxJDWoGnI7D25BY/D09VKyn6sl4PDYeyVKulSSZ6RSOqqaXVytVxGs40momAegAJLC6TiRMZ7dT3LDGQa2gB2dBFfrkVE7fWYJ6cBjS4o8b7oUgFIqlQy4uEAeTDLAj2NjDRKgEB/lH3NFPHnvT5RmPyxrNIxCPGwstxiu5ujdehIM4XYZWazoEBXNSaBJ3IPo9Ch8OySNYX7/HGmEFg2KQ3AeGGw9bWPNDp6Tgk0I5d2egpcLqEr/G7pjrps9Vsc9vrbY9679qyDqrD0dp8dTbfOoEH+cQieIirlsG5vlucAYGeRRMNs/5HoBS7QsB0FHMoYGPCOqbpqWmbxlWNYpmOkb3tGeGNI2+bBlhxHfpBZHlk0CbVsm2Ffh2LgMfWTSUUg7hAA===

Highlight here:
```
    [CompilerGenerated]
    public override string ToString()
    {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.Append("ModelA");
        stringBuilder.Append(" { ");
        if (PrintMembers(stringBuilder))
        {
            stringBuilder.Append(' ');
        }
        stringBuilder.Append('}');
        return stringBuilder.ToString();
    }
```

`sealed override` does work, see:
https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMAWAbj0JMwDoAVAUwA9grc8A3AQzCIC2AezhMANgGciAXiJQmAdyIAZAJYTgAHgBCPCUwCyI8QD4AFAEoiAbzmKiR0WICClmHaWPx2t0QC+HHgAZkJgTDwAxgAWRGa8/MJORKpQgsaSFnjWeES5tACcZonijEIAysBgKQDmlhYceQVF6aUVVVDV6HUNeaSF9Xh+eHhMUBACDukMAJ4ADkw2RACqUADWUEIKsJNOzu5eYtr+eDwARhpgkcBEYRGhcES6+gdZOb0AzERcqmDAEDxiWjEABKPAUM3mi2qTHYySgqlhfhkRAARCt1psoCierlUJ8DhCFoSZCYiABRMYCOgABT4+k0BLmTHMoPBTIGuEaszAQmATAifIe31+/0BpyEQkB1PawAMTAEpyYYAkZjaNW0EFUYlE/FOmu1SsynLy2WNjVyeq1Orozlm8ygcDMABIUazibJrG6mX4URzzb0AOxESoQJg445mkiffQApgPIRcJVVURAogMcqVGqWEmogDyibAyaJGfa1UAgP/Yt64z7Cv4A1PptUdLpWaSklGNzPNiscIacXC3e47cTOIggR56QzpV6RvFEBNJ1Qp0ggsHE2zQ2EpBEUfzIlEHZyV2efFfktiXUcbmG77eI/dki88I+94YD/lDg5HcdPKdOGeNHOC6FkuCxnl6kLXlu8L3rIB7pNox6AaemDEI+lQ8EcUG3jBu5InB6GXIh4YAPQkVG84FkWDYllmrbtvmi4pl+aa0R0PZVhRwHUWenali2OYooxIHMQhrFNp0HG4H4QA==

**Comment by bunnyi116** [comment]
> > `ModelA` failed to get the `override string ToString()` from `BaseModel`. It used `record`s default `ToString()`, which means `BaseModel`, `override string ToString()` does not work.
> 
> If you decompile the sample, you will see that `ModelA` overrides it again. There's not a "default" implementation for `record`. Every record just reimplement it in that way. See: https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwCwG5lqZwB0AKgKYAeALvksgG4CGATugLYD2w5ANgM7oAvOgB25AO7oAMgEt+1ADwAhJv3IBZbnwB8ACgCU6AN5jJ6LT14BBQwBozUy32X30AXzrIAZpxbkmAGMAC3Q9ZjYuK3RZUQ5tAQNkY2R0NKIATj0ovjJOAGVqFliAc0MDOnTM7IS8wuLREphyyvSsLIrkd2RkclEAV3YLBNIATwAHchN0AFVRAGtRTglRB2cbNYTlD2QmACMFFiDqdH9Av2B0VXV15NS2gGZ0BlkWan6mXiIMACUmCTGk2mJXItBiolkYPcwnQACI5otlqJYa00lAnutAVMscIdOgAKIDdjEAAKrHUikxE3I+j+AOpnSQVXGLE41HIgXZlxebw+Xz2nE4XxJDWoGnI7D25BY/D09VKyn6sl4PDYeyVKulSSZ6RSOqqaXVytVxGs40momAegAJLC6TiRMZ7dT3LDGQa2gB2dBFfrkVE7fWYJ6cBjS4o8b7oUgFIqlQy4uEAeTDLAj2NjDRKgEB/lH3NFPHnvT5RmPyxrNIxCPGwstxiu5ujdehIM4XYZWazoEBXNSaBJ3IPo9Ch8OySNYX7/HGmEFg2KQ3AeGGw9bWPNDp6Tgk0I5d2egpcLqEr/G7pjrps9Vsc9vrbY9679qyDqrD0dp8dTbfOoEH+cQieIirlsG5vlucAYGeRRMNs/5HoBS7QsB0FHMoYGPCOqbpqWmbxlWNYpmOkb3tGeGNI2+bBlhxHfpBZHlk0CbVsm2Ffh2LgMfWTSUUg7hAA===
> 
> Highlight here:
> 
> ```
>     [CompilerGenerated]
>     public override string ToString()
>     {
>         StringBuilder stringBuilder = new StringBuilder();
>         stringBuilder.Append("ModelA");
>         stringBuilder.Append(" { ");
>         if (PrintMembers(stringBuilder))
>         {
>             stringBuilder.Append(' ');
>         }
>         stringBuilder.Append('}');
>         return stringBuilder.ToString();
>     }
> ```
> 
> `sealed override` does work, see: https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMAWAbj0JMwDoAVAUwA9grc8A3AQzCIC2AezhMANgGciAXiJQmAdyIAZAJYTgAHgBCPCUwCyI8QD4AFAEoiAbzmKiR0WICClmHaWPx2t0QC+HHgAZkJgTDwAxgAWRGa8/MJORKpQgsaSFnjWeES5tACcZonijEIAysBgKQDmlhYceQVF6aUVVVDV6HUNeaSF9Xh+eHhMUBACDukMAJ4ADkw2RACqUADWUEIKsJNOzu5eYtr+eDwARhpgkcBEYRGhcES6+gdZOb0AzERcqmDAEDxiWjEABKPAUM3mi2qTHYySgqlhfhkRAARCt1psoCierlUJ8DhCFoSZCYiABRMYCOgABT4+k0BLmTHMoPBTIGuEaszAQmATAifIe31+/0BpyEQkB1PawAMTAEpyYYAkZjaNW0EFUYlE/FOmu1SsynLy2WNjVyeq1Orozlm8ygcDMABIUazibJrG6mX4URzzb0AOxESoQJg445mkiffQApgPIRcJVVURAogMcqVGqWEmogDyibAyaJGfa1UAgP/Yt64z7Cv4A1PptUdLpWaSklGNzPNiscIacXC3e47cTOIggR56QzpV6RvFEBNJ1Qp0ggsHE2zQ2EpBEUfzIlEHZyV2efFfktiXUcbmG77eI/dki88I+94YD/lDg5HcdPKdOGeNHOC6FkuCxnl6kLXlu8L3rIB7pNox6AaemDEI+lQ8EcUG3jBu5InB6GXIh4YAPQkVG84FkWDYllmrbtvmi4pl+aa0R0PZVhRwHUWenali2OYooxIHMQhrFNp0HG4H4QA==

But then derived classes cannot override ToString

**Comment by Joe4evr** [comment]
> But then derived classes cannot override ToString

![](https://i.imgur.com/jks6Qqr.png)

**Comment by ericstj** [comment]
Moving this to Roslyn since this issue is about what the compiler emits for derived record types.  It sounds to me like @Bunnui is requesting some way to tell the compiler not to implement ToString on the derived type.

**Comment by CyrusNajmabadi** [comment]
Moving to csharplang.  Roslyn is emitting correctly as per the specification afaict.  This is a request that the specification around records either be changed, or have some update to it to allow customization of this behavior.  Our currently position though is that we are highly unlikely to make any changes here.  Instead, you can fully customize the behavior of records by supplying custom implementations yourself.

And, for something like:

> For the PrintMemberAttribute suggestion, the attributes can be: IgnorePrintMembers, IgnoreNull

This is already possible today if someone wants to write a source-generator.  Just define those attributes, and write a source-generator that spits out the methods you want that behave with the semantics you want those attributes to have.
